### PR TITLE
Added Waypoints widget

### PR DESCRIPTION
**Pull request by nma83**
Owntracks Waypoint widget is an Android home-screen widget. It uses 2 pieces of information from each user who has shared his/her location:
- their current location
- their shared waypoints

The widget compares the location with all possible pairs of waypoints and (possibly) comes up with a pair that bounds the current location best. The waypoints from this pair are shown on the ends of a linear scale and the current location is placed on the scale based on distance from each end.
More details at: http://narenma.net/2016/07/30/owntracks-waypoints-widget/

I will try to walk-through the code changes file-by-file (most critical first):
- App.java
  - added a HashMap to store (FusedContact, WaypointCollection) for all contacts
  - hooks to update the widget on location updates
- ServiceMessage.java
  - process a waypoints message and update list of waypoints for a contact
- **NEW** ServiceWidget.java
  - service added to provide data to the widget views
  - most the widget logic is contained in this service
- **NEW** WaypointPair.java
  - helper class to compute bounds and select waypoint pairs to display
- **NEW** LClocWidgetProvider.java
  - class added to manage the widget
- ContactImageProvider.java
  - had to change the visibility of a few helper static methods to public to access them from outside for the widget views
- ServiceApplication.java
  - added a method to request other users' waypoint using a MessageCmd, but it is not working as intended due to the way processMessage is coded in ServiceBroker (but can look at it later)
- Preferences are updated to add a new `Widget` item under `Notifications`
- All required resources are added into the appropriate res sub-directories
- Had to change gradle version to get past a dependency fetch error I ran into

Please let me know if any more details are required.

Thanks!
Narendra

### CODE REVIEW

**Comment by binarybucks** [comment]
Thank you for your contribution and the great documentation. 
As I'm currently working on Android 7 support on a master branch, I'll leave this pull-request open and merge the changes afterwards, as the Android 7 branch probably contains some breaking changes.

**Comment by binarybucks** [comment]
One thing that occured to me though: You're loading the waypoints on the main thread when the app starts. In the Andorid 7 branch I moved the service initialization to a backround thread to increase the app startup speed. I think we should do this here too. 

It's probably an easy fix by moving the loading code to a runnable.

**Comment by nma83** [comment]
Sure, will change the waypoint loading and add the commit to this pull request.

**Comment by nma83** [comment]
When you say waypoint loading, do you mean requestWaypoints()? Or is it updateWidget()?
requestWaypoints() is already run in a Runnable.

**Comment by nma83** [comment]
@binarybucks do I need to rebase and resolve for this PR?

**Comment by nma83** [comment]
I am having a really tough time trying to rebase my changes on top of `master`. Have to try manual merging :(.

File: project/app/src/main/java/org/owntracks/android/services/ServiceMessageHttp.java [hunk 1]
```diff
@@ -14,7 +14,7 @@
 import com.google.android.gms.gcm.OneoffTask;
 import com.google.android.gms.gcm.Task;
 
-import org.antlr.v4.runtime.misc.NotNull;
```
Comment (nma83) [inline project/app/src/main/java/org/owntracks/android/services/ServiceMessageHttp.java]: This package was flagged as non-existing in my build, even with just the baseline upstream/master repo without my changes.

File: project/app/src/main/java/org/owntracks/android/services/ServiceMessageHttp.java [hunk 2]
```diff
@@ -275,7 +275,7 @@ public static int postMessage(final String body, @Nullable final String url, @Nu
     }
 
 
-    private static int onMessageDelivered(@NotNull Context c, @Nullable Long messageId) {
```
Comment (nma83) [inline project/app/src/main/java/org/owntracks/android/services/ServiceMessageHttp.java]: Does not seem to exist in latest antlr.

**Comment by nma83** [comment]
Finally managed to rebase properly (I hope). @binarybucks, is it possible to get some feedback on the changes? There is a build issue with antlr that I have noted in the line comments.

**Comment by nma83** [comment]
@binarybucks is there any feedback on this PR?

**Pull request closed without merging**